### PR TITLE
Bug Fix: Adding back the plugin architecture logic for talawa app 

### LINKED ISSUE #1758: Bug Report: Adding back the plugin architecture logic for talawa app
**Describe the bug**
Recently, due to a code commenting error, some part of the code has been deleted from talawa, talawa-admin and talawa-api.
The reason for this being contributors didn't read the [docs on Plugins](https://docs.talawa.io/docs/category/plugins) before deleting those files and comments from the code. That code needs to be re-added and made supportive of the current major changes in the project to make it work again.
 
**To Reproduce**
Steps to reproduce the behaviour:

1. Try installing plugins on talawa-admin
2. See changes in the app
3. nothing happens

**Expected behaviour**
Plugins should work fine means app features should be controlled by talawa-admin

**Actual behaviour**
Nothing happens after uninstalling or installing plugins on the app.

**Screenshots**
N/A

**Additional details**
N/A

### PR DESCRIPTION
**Pull request by SiddheshKukade**
**What kind of change does this PR introduce?**
Refactoring

**Issue Number:**

Fixes #1758 

**Did you add tests for your changes?**
Not required.
**Snapshots/Videos:**

block diagram for overview: 
image

[Watch the video](https://user-images.githubusercontent.com/65951872/235835450-7a248f03-f235-4a8c-852d-49d6d3428b41.mp4)

**If relevant, did you update the documentation?**
To make it work with mongoDB ,talawa-api , talawa-admin we have to firstly store some data in the db and then we can use the store further 
The `plugins` collection should have follwing with proper orgIDs for testing puroposes:
```
 {
        "_id": "63929c1915ad470d13c3bcb6",
        "pluginInstallStatus": true,
        "pluginName": "Events",
        "pluginCreatedBy": " Mr. Siddhesh Bhupendra  Kukade",
        "pluginDesc":
            "Presenting the Events Plugin by Siddhesh Bhupendra Kuakde that implements the plugin features as a component here members can create and manage events and on top of it everything is controlled by the Organization Admins . Install and use right away.",
        "installedOrgs": [
          "62ccfccd3eb7fd2a30f41601",
          "62ccfccd3eb7fd2a30f41601",
          "<YOUR_ORG_ID_FOR_DEVELOPMENT>"
        ]
      },
      {
        "_id": "63929c1915ad470d13c3bcb6",
        "pluginInstallStatus": true,
        "pluginName": "Posts",
        "pluginCreatedBy": "Mr.Johnathan Doe ",
        "pluginDesc":
            "Presenting the Donation plugin for the talawa app that enables members of the organization to  do One time or Reccurinng donations to an  organization",
        "installedOrgs": [
          "62ccfccd3eb7fd2a30f41601",
          "62ccfccd3eb7fd2a30f41601",
          "<YOUR_ORG_ID_FOR_DEVELOPMENT>"
        ]
      },
      {
        "_id": "62cfcf71824dc26e1fbeed46",
        "pluginInstallStatus": true,
        "pluginName": "Donation",
        "pluginCreatedBy": "Kimia",
        "pluginDesc":
            "See all available organizations in entire world using this plugins",
        "installedOrgs": [
          "62ccfccd3eb7fd2a30f41601",
          "<YOUR_ORG_ID_FOR_DEVELOPMENT>"
        ]
      },
      {
        "_id": "62cfcf71824dc26e1fbeed46",
        "pluginInstallStatus": true,
        "pluginName": "Chats",
        "pluginCreatedBy": "Kimia",
        "pluginDesc":
            "See all available organizations in entire world using this plugins",
        "installedOrgs": [
          "62ccfccd3eb7fd2a30f41601",
          "<YOUR_ORG_ID_FOR_DEVELOPMENT>"
        ]
      }
```

**Summary**
 Added the 15 sec delay to plugin calls and added back the previous deletions.

**Does this PR introduce a breaking change?**
No

 

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa/blob/master/CONTRIBUTING.md)?**

Yes

### CODE REVIEW

**Comment by literalEval** [comment]
@palisadoes @SiddheshKukade  please tag me once this is ready for review. Since I worked on refactoring, I can advise a thing or two in this PR. Thanks.

**Comment by palisadoes** [comment]
Thanks for submitting your PR, but we must close it.

1. We get notified of every event in the queue, and draft PRs like these, and their updates that require no action can be a distraction. We need our reviewers to focus on updates that are ready. 
1.  Don't be discouraged, we are just trying to keep our workloads to a manageable level.

Please resubmit when you are ready.

**Comment by SiddheshKukade** [comment]
@palisadoes Sir, sorry I wasn't aware about the notification thing.
I just opened the PR as draft to check for any CI/CD error that might occur.

It's alright, I'll figure out something else to check that 👍

**Comment by SiddheshKukade** [comment]
@palisadoes, Sir the code for this is ready reviewed, please re-open this PR to see the latest commits.

File: lib/views/main_screen.dart
Comment (literalEval) [inline lib/views/main_screen.dart:219]: This violates the DRY principle. Have a look at the refactored code and how it was converted into a function to avoid repetition.

File: lib/views/main_screen.dart
Comment (literalEval) [inline lib/views/main_screen.dart:54]: Having demo data in the widget class is not a good idea. Having it in a separate `data` file will make the code better.

File: lib/main.dart
Comment (literalEval) [inline lib/main.dart:187]: This will entirely break the app.

1. There is no need to create a system-wide timer. It can be local to the MainScreen.
2. You are fetching plugins unconditionally. This will cause a null exception when the user has not entered org URL yet. There was a bugfix regarding this sometime ago. This code change will reintroduce that bug.

File: lib/views/main_screen.dart
Comment (literalEval) [inline lib/views/main_screen.dart:145]: Having static data in the build method is not a good idea. Please consider refactoring this.

**Comment by literalEval** [comment]
@SiddheshKukade please have a look at the previously refactored code and bug fixes thoroughly, so that we don't reinvent the wheel :)

**Review by literalEval — CHANGES_REQUESTED** [review]

File: lib/views/main_screen.dart
Comment (SiddheshKukade) [reply lib/views/main_screen.dart:145]: That's how we planned to do it, previously

File: lib/main.dart
Comment (SiddheshKukade) [reply lib/main.dart:187]: Users can't enter without having the org URL.

Comment (SiddheshKukade) [reply lib/main.dart:187]: Nothing is getting "breaked" on my side.

File: lib/views/main_screen.dart
Comment (SiddheshKukade) [reply lib/views/main_screen.dart:219]: The current code is not compatible with the plugin logic.

File: lib/views/main_screen.dart
Comment (SiddheshKukade) [reply lib/views/main_screen.dart:54]: The main focus behind the PR was to bring back the code which was previously deleted.

File: lib/views/main_screen.dart
Comment (literalEval) [reply lib/views/main_screen.dart:219]: It is. Have a deeper thought over it. And if you still can't figure it out, then maybe change the logic a bit.

File: lib/main.dart
Comment (literalEval) [reply lib/main.dart:187]: > Nothing is getting "breaked" on my side.

That is because you have not performed an E2E testing after clearing the data.

File: lib/views/main_screen.dart
Comment (literalEval) [reply lib/views/main_screen.dart:145]: We can improve now

**Comment by literalEval** [comment]
@SiddheshKukade don't mark reviews as resolved without the mentors and admins commenting on it.

**Comment by SiddheshKukade** [comment]
@DMills27 Sir, I would also like to your thoughts on this.

I've also updated the description with more details.

**Comment by palisadoes** [comment]
@SiddheshKukade please get your mentors to review and approve this.

**Comment by SiddheshKukade** [comment]
> @SiddheshKukade please get your mentors to review and approve this.

@palisadoes Sure, Actually, some tasks were pending before this merged to avoid inconsistencies. I'll link them here 👍🏻

**Comment by literalEval** [comment]
@palisadoes Since the plugin system is a GSoC project anyway, and since no project directly depends on this, I think instead of reintroducing the previous work, we can wait till @SiddheshKukade fixes everything this time, and close this PR without a merge.

**Comment by SiddheshKukade** [comment]
@palisadoes sir, I'm having practical exams from the past 2 weeks. Due to that I had to take a break untill 3rd june. I have also discussed that with my mentor tasneem mam.
My other PRs need to be merged before this get merged to avoid inconsistencies.

I'll get started to complete that In a few days and there is no need to close the PR.
Thanks

**Comment by palisadoes** [comment]
@SiddheshKukade What should we do about this PR?

**Comment by palisadoes** [comment]
@tasneemkoushar @DMills27 Please add your comments

**Comment by SiddheshKukade** [comment]
> @SiddheshKukade What should we do about this PR?

@palisadoes Sir, currently this should PR be kept on hold because merging this before resolving https://github.com/PalisadoesFoundation/talawa-api/issues/1337 can create problem for other contributors for developing the app without having the basic plugin data in their database. Please wait until the PR for api#1337 is merged.

**Comment by SiddheshKukade** [comment]
**Update**: I was thinking about adding the websockets logic (which is next step in my project ) in this PR
This will ensure that other contributors would not face any problems with their work regarding plugins. 

So for that reason this has to be kept open for a few more days.

**Review by noman2002 — COMMENTED** [review]
Can you Add a video of showing what will happen after 15 secs if we disable/enable a feature from admin.

File: lib/plugins/fetch_plugin_list.dart
Comment (SiddheshKukade) [reply lib/plugins/fetch_plugin_list.dart:11]: Yes, actually I need to add the socket listeners ( GraphQL Subscriptions) .

**Pull request closed without merging**
